src/main/java/io/debezium/connector/vitess/VitessConnector.java

```diff
@@ -123,23 +131,30 @@ public List<Map<String, String>> taskConfigs(int maxTasks, List<String> currentS
         if (connectorConfig.offsetStoragePerTask()) {
             final int prevNumTasks = connectorConfig.getPrevNumTasks();
             final int gen = connectorConfig.getOffsetStorageTaskKeyGen();
-            int tasks = Math.min(maxTasks, currentShards.size());
+            int tasks = Math.min(maxTasks, currentShards == null ? Integer.MAX_VALUE : currentShards.size());
             LOGGER.info("There are {} vitess shards for maxTasks: {}, we will use {} tasks",
-                    currentShards.size(), maxTasks, tasks);
-            if (gen > 0 && tasks == prevNumTasks) {
+                    currentShards == null ? null : currentShards.size(), maxTasks, tasks);
+            // Check the task offsets persisted from previous gen, we expect the offsets are saved
+            Map<String, String> prevGtidsPerShard = gen > 0 ? getGtidPerShardFromStorage(prevNumTasks, gen - 1, true) : null;
+            LOGGER.info("Previous gtids Per shard: {}", prevGtidsPerShard);
+            Set<String> previousShards = prevGtidsPerShard != null ? prevGtidsPerShard.keySet() : null;
+            if (gen > 0 && tasks == prevNumTasks && hasSameShards(previousShards, currentShards)) {
                 throw new IllegalArgumentException(String.format(
                         "Previous num.tasks: %s and current num.tasks: %s are the same. "
+                                + "And previous shards: %s and current shards: %s are the same. "
                                 + "Please choose different tasks.max or have different number of vitess shards "
                                 + "if you want to change the task parallelism.  "
                                 + "Otherwise please reset the offset.storage.task.key.gen config to its original value",
-                        prevNumTasks, tasks));
+                        prevNumTasks, tasks, previousShards, currentShards));
             }
-            // Check the task offsets persisted from previous gen, we expect the offsets are saved
-            Map<String, String> prevGtidsPerShard = gen > 0 ? getGtidPerShardFromStorage(prevNumTasks, gen - 1, true) : null;
-            LOGGER.info("Previous gtids Per shard: {}", prevGtidsPerShard);
             if (prevGtidsPerShard != null && !hasSameShards(prevGtidsPerShard.keySet(), currentShards)) {
                 LOGGER.warn("Some shards for the previous generation {} are not persisted.  Expected shards: {}",
                         prevGtidsPerShard.keySet(), currentShards);
+                if (prevGtidsPerShard.keySet().containsAll(currentShards)) {
```

Comment:
Should we throw Exception in this case?   People usually don't look at the logs until problem happens.    If we let this pass, we won't have the old gtids for shrinked shards anymore since we don't look for older gtids more than one generation old.   If we think shrinking shards is a valid use case, then we probably need to a config flag to indicate whether we should halt on shard list shrinking or not, the default value for that config value should be false.

Reply:
Thanks for the update, it looks good to me now.

Reply:
Thanks for the feedback. Updated to throw an exception.

If we end up with a use case later, we can implement functionality for contracting the shard list, but for now we will opt for the more cautious path of preventing any lost state (plus this is a new feature, so there's no use cases that do this contracting shard list, and we only know of a use case for expansion i.e., ours).